files/upload_sig.cgi

```diff
@@ -0,0 +1,282 @@
+
+
+def send_email(pkg, sha1, filename, username, branch=''):
+    text = """A file has been added to the lookaside cache for %(pkg)s:
+
+%(branch)s %(sha1)s  %(filename)s""" % locals()
+    msg = MIMEText(text)
+    sender_name = conf.get('mail', 'sender_name')
+    sender_email = conf.get('mail', 'sender_email')
+    sender = Utils.formataddr((sender_name, sender_email))
+    recipient = conf.get('mail', 'recipient')
+    msg['Subject'] = 'File %s uploaded to lookaside cache, branch %s by %s' % (
+            filename, branch, username)
+    msg['From'] = sender
+    msg['To'] = recipient
+    try:
+        s = smtplib.SMTP(conf.get('mail', 'smtp_server'))
+        s.sendmail(sender, recipient, msg.as_string())
+    except:
+        errstr = 'sending mail for upload of %s failed!' % filename
+        print >> sys.stderr, errstr
+        syslog.syslog(errstr)
+
+
+def main():
+    os.umask(002)
+
+    username = os.environ.get('SSL_CLIENT_S_DN_CN', None)
+
+    assert os.environ['REQUEST_URI'].split('/')[1] == 'sources'
+
+    form = cgi.FieldStorage()
+    name = check_form(form, 'name')
+    branch = check_form(form, 'branch', exit_on_error=False)
+    is_hash_format = branch is None
+
+    # Search for the file hash, start with stronger hash functions
+    if 'sha512sum' in form:
+        checksum = check_form(form, 'sha512sum')
+        hash_type = "sha512"
+
+    elif 'md5sum' in form:
+        # Fallback on md5, as it's what we currently use
+        checksum = check_form(form, 'md5sum')
+        hash_type = "md5"
+
+    elif 'sha1sum' in form:
+        checksum = check_form(form, 'sha1sum')
+        hash_type = "sha1"
+    else:
+        send_error('Required checksum is not present.')
+
+    if branch is None and not is_hash_format:
+        send_error('Couldn\'t identify cache format.')
+
+    action = None
+    upload_file = None
+    filename = None
+
+    # Is this a submission or a test?
+    # in a test, we don't get a file.
+    if 'file' not in form:
+        action = 'check'
+        if is_hash_format:
+            print >> sys.stderr, '[username=%s] Checking file status: NAME=%s HASH=%s SHA1SUM=%s' % (username, name, hash_type, checksum)
+        else:
+            print >> sys.stderr, '[username=%s] Checking file status: NAME=%s BRANCH=%s SHA1SUM=%s' % (username, name, branch, checksum)
+    else:
+        action = 'upload'
+        upload_file = form['file']
+        if not upload_file.file:
+            send_error('No file given for upload. Aborting.')
+        filename = os.path.basename(upload_file.filename)
+        if is_hash_format:
+            print >> sys.stderr, '[username=%s] Processing upload request: NAME=%s HASH=%s CHECKSUM=%s' % (username, name, hash_type, checksum)
+        else:
+            print >> sys.stderr, '[username=%s] Processing upload request: NAME=%s BRANCH=%s CHECKSUM=%s' % (username, name, branch, checksum)
+
+    if is_hash_format:
+        module_dir_args = [name, filename, hash_type, checksum]
```

Comment:
I am use `filename` as the tarball name in here, is it ok?

Reply:
This looks right